File: src/redact.ts
import type {
  PlaceholderContext,
  PlaceholderFormatter,
  RedactOptions,
  SecretAction,
  SecretConfidence,
  SecretFinding,
} from "./types.js";

export type SecretRedactionErrorCode =
  | "INVALID_INPUT"
  | "INVALID_FINDINGS"
  | "INVALID_OPTIONS"
  | "PLACEHOLDER_FAILURE"
  | "INVALID_PLACEHOLDER";

const ERROR_MESSAGES: Readonly<Record<SecretRedactionErrorCode, string>> = {
  INVALID_INPUT: "Secret redaction input must be a string.",
  INVALID_FINDINGS: "Secret redaction findings are invalid.",
  INVALID_OPTIONS: "Secret redaction options are invalid.",
  PLACEHOLDER_FAILURE: "The placeholder formatter failed.",
  INVALID_PLACEHOLDER: "The placeholder formatter returned an invalid value.",
};

const IDENTIFIER_PATTERN = /^[a-z][a-z0-9]*(?:[._-][a-z0-9]+)*$/;
const MAX_IDENTIFIER_LENGTH = 64;
const MAX_PLACEHOLDER_LENGTH = 256;

interface ForbiddenMatchedTextIndex {
  readonly byLength: ReadonlyMap<number, ReadonlySet<string>>;
  readonly lengths: readonly number[];
}

export class SecretRedactionError extends Error {
  readonly code: SecretRedactionErrorCode;

  constructor(code: SecretRedactionErrorCode) {
    super(ERROR_MESSAGES[code]);
    this.name = "SecretRedactionError";
    this.code = code;
  }
}

function isIdentifier(value: unknown): value is string {
  return (
    typeof value === "string" &&
    value.length <= MAX_IDENTIFIER_LENGTH &&
    IDENTIFIER_PATTERN.test(value)
  );
}

function isConfidence(value: unknown): value is SecretConfidence {
  return value === "high" || value === "medium" || value === "low";
}

function isAction(value: unknown): value is SecretAction {
  return (
    value === "redact" ||
    value === "block" ||
    value === "warn" ||
    value === "allow"
  );
}

function normalizeFinding(input: string, value: unknown): SecretFinding {
  if (typeof value !== "object" || value === null) {
    throw new SecretRedactionError("INVALID_FINDINGS");
  }

  try {
    const finding = value as Partial<SecretFinding>;
    if (
      !isIdentifier(finding.id) ||
      !isIdentifier(finding.type) ||
      !isIdentifier(finding.detector) ||
      !isConfidence(finding.confidence) ||
      !isAction(finding.action) ||
      !Number.isInteger(finding.start) ||
      !Number.isInteger(finding.end) ||
      finding.start === undefined ||
      finding.end === undefined ||
      finding.start < 0 ||
      finding.end <= finding.start ||
      finding.end > input.length
    ) {
      throw new TypeError();
    }

    return Object.freeze({
      id: finding.id,
      type: finding.type,
      detector: finding.detector,
      confidence: finding.confidence,
      action: finding.action,
      start: finding.start,
      end: finding.end,
    });
  } catch {
    throw new SecretRedactionError("INVALID_FINDINGS");
  }
}

export const defaultPlaceholderFormatter: PlaceholderFormatter = (
  _finding,
  context,
) => `<SECRET_${context.placeholderIndex}>`;

export const typedPlaceholderFormatter: PlaceholderFormatter = (
  finding,
  context,
) => `<${finding.type.toUpperCase().replace(/[.-]/g, "_")}_${context.placeholderIndex}>`;

function resolveFormatter(options: RedactOptions): PlaceholderFormatter {
  if (typeof options !== "object" || options === null) {
    throw new SecretRedactionError("INVALID_OPTIONS");
  }
  try {
    const formatter =
      options.placeholderFormatter ?? defaultPlaceholderFormatter;
    if (typeof formatter !== "function") throw new TypeError();
    return formatter;
  } catch {
    throw new SecretRedactionError("INVALID_OPTIONS");
  }
}

function formatPlaceholder(
  formatter: PlaceholderFormatter,
  finding: SecretFinding,
  context: PlaceholderContext,
  forbiddenMatchedTexts: ForbiddenMatchedTextIndex,
): string {
  let placeholder: unknown;
  try {
    placeholder = formatter(finding, context);
  } catch {
    throw new SecretRedactionError("PLACEHOLDER_FAILURE");
  }

  if (
    typeof placeholder !== "string" ||
    placeholder.length === 0 ||
    placeholder.length > MAX_PLACEHOLDER_LENGTH ||
    containsForbiddenMatchedText(placeholder, forbiddenMatchedTexts)
  ) {
    throw new SecretRedactionError("INVALID_PLACEHOLDER");
  }
  return placeholder;
}

function buildForbiddenMatchedTextIndex(
  input: string,
  findings: readonly SecretFinding[],
): ForbiddenMatchedTextIndex {
  const byLength = new Map<number, Set<string>>();

  for (const finding of findings) {
    if (finding.action !== "redact" && finding.action !== "block") continue;

    const length = finding.end - finding.start;
    if (length > MAX_PLACEHOLDER_LENGTH) {
      continue;
    }

    let values = byLength.get(length);
    if (values === undefined) {
      values = new Set<string>();
      byLength.set(length, values);
    }
    values.add(input.slice(finding.start, finding.end));
  }

  return {
    byLength,
    lengths: [...byLength.keys()].sort((left, right) => left - right),
  };
}

function containsForbiddenMatchedText(
  placeholder: string,
  index: ForbiddenMatchedTextIndex,
): boolean {
  for (const length of index.lengths) {
    if (length > placeholder.length) break;

    const forbiddenValues = index.byLength.get(length);
    if (forbiddenValues === undefined) continue;

    for (let start = 0; start + length <= placeholder.length; start += 1) {
      if (forbiddenValues.has(placeholder.slice(start, start + length))) {
        return true;
      }
    }
  }
  return false;
}

/**
 * Reconstructs text in one pass. Findings with `warn` or `allow` remain
 * unchanged; `redact` and `block` ranges receive deterministic placeholders.
 */
export function redact(
  input: string,
  findings: readonly SecretFinding[],
  options: RedactOptions = {},
): string {
  if (typeof input !== "string") {
    throw new SecretRedactionError("INVALID_INPUT");
  }
  if (!Array.isArray(findings)) {
    throw new SecretRedactionError("INVALID_FINDINGS");
  }

  const formatter = resolveFormatter(options);
  let normalized: SecretFinding[];
  try {
    normalized = findings
      .map((finding) => normalizeFinding(input, finding))
      .sort((left, right) => left.start - right.start || left.end - right.end);
  } catch {
    throw new SecretRedactionError("INVALID_FINDINGS");
  }

  let previousEnd = 0;
  for (const finding of normalized) {
    if (finding.start < previousEnd) {
      throw new SecretRedactionError("INVALID_FINDINGS");
    }
    previousEnd = finding.end;
  }

  const chunks: string[] = [];
  let cursor = 0;
  let placeholderIndex = 0;
  const forbiddenMatchedTexts = buildForbiddenMatchedTextIndex(
    input,
    normalized,
  );
  for (const finding of normalized) {
    if (finding.action === "warn" || finding.action === "allow") continue;

    chunks.push(input.slice(cursor, finding.start));
    placeholderIndex += 1;
    const context = Object.freeze({ placeholderIndex });
    chunks.push(
      formatPlaceholder(
        formatter,
        finding,
        context,
        forbiddenMatchedTexts,
      ),
    );
    cursor = finding.end;
  }
  chunks.push(input.slice(cursor));
  return chunks.join("");
}

File: crates/secret-scan-core/src/redact.rs
//! Single-pass redaction: replaces `block` and `redact` findings with
//! placeholder text produced by a [`PlaceholderFormatter`], leaving `warn`
//! and `allow` findings untouched.
//!
//! Redaction never exposes a matched value: it does not accept the value
//! itself (only [`Finding`] metadata and the immutable `input`), and every
//! failure is a sanitized [`SecretScanError`].

use std::collections::{BTreeMap, BTreeSet};

use crate::error::{FormatterFailure, SecretScanError, SecretScanErrorCode};
use crate::types::{ByteRange, Finding, PlaceholderContext, PlaceholderFormatter};

/// Maximum length in bytes of a placeholder a [`PlaceholderFormatter`] may
/// return.
pub const MAX_PLACEHOLDER_LENGTH: usize = 256;

/// The default formatter: `<SECRET_N>`, `N` one-based among replaced
/// findings.
///
/// # Errors
///
/// Never fails.
pub fn default_placeholder_formatter(
    _finding: &Finding,
    context: &PlaceholderContext,
) -> Result<String, FormatterFailure> {
    Ok(format!("<SECRET_{}>", context.placeholder_index()))
}

/// A formatter that names the finding type: `<TYPE_NAME_N>`, upper-cased
/// with `.` and `-` mapped to `_`.
///
/// # Errors
///
/// Never fails.
pub fn typed_placeholder_formatter(
    finding: &Finding,
    context: &PlaceholderContext,
) -> Result<String, FormatterFailure> {
    let normalized = finding
        .type_name()
        .to_ascii_uppercase()
        .replace(['.', '-'], "_");
    Ok(format!("<{normalized}_{}>", context.placeholder_index()))
}

/// Matched values eligible to be reproduced by a placeholder, indexed by
/// byte length so a formatter output can be checked in one pass over its
/// substrings.
///
/// Only `redact`/`block` findings whose matched text is no longer than
/// [`MAX_PLACEHOLDER_LENGTH`] are eligible: a longer matched value can never
/// fit inside a valid placeholder, so indexing it would be wasted work.
struct ForbiddenMatchedText {
    by_length: BTreeMap<usize, BTreeSet<String>>,
}

impl ForbiddenMatchedText {
    fn build(input: &str, findings: &[&Finding]) -> Self {
        let mut by_length: BTreeMap<usize, BTreeSet<String>> = BTreeMap::new();
        for finding in findings {
            if !finding.action().replaces_text() {
                continue;
            }
            let range = finding.range();
            if range.len() > MAX_PLACEHOLDER_LENGTH {
                continue;
            }
            by_length
                .entry(range.len())
                .or_default()
                .insert(input[range.start()..range.end()].to_string());
        }
        Self { by_length }
    }

    /// `true` when `placeholder` contains any indexed matched value as a
    /// substring.
    ///
    /// `by_length` iterates in ascending key order, so the scan stops as
    /// soon as a length exceeds what remains of `placeholder`.
    fn contains(&self, placeholder: &str) -> bool {
        for (&length, values) in &self.by_length {
            if length == 0 || length > placeholder.len() {
                continue;
            }
            let mut start = 0;
            while start + length <= placeholder.len() {
                if placeholder.is_char_boundary(start)
                    && placeholder.is_char_boundary(start + length)
                    && values.contains(&placeholder[start..start + length])
                {
                    return true;
                }
                start += 1;
            }
        }
        false
    }
}

fn ordered_and_disjoint<'a>(
    findings: &'a [Finding],
    input: &str,
) -> Result<Vec<&'a Finding>, SecretScanError> {
    for finding in findings {
        if !finding.range().is_char_aligned_in(input) {
            return Err(SecretScanErrorCode::InvalidFindings.into());
        }
    }

    let mut ordered: Vec<&Finding> = findings.iter().collect();
    ordered.sort_by_key(|finding| (finding.range().start(), finding.range().end()));

    let mut previous_end = 0;
    for finding in &ordered {
        let range: ByteRange = finding.range();
        if range.start() < previous_end {
            return Err(SecretScanErrorCode::InvalidFindings.into());
        }
        previous_end = range.end();
    }

    Ok(ordered)
}

/// Reconstructs `input` in one ordered pass: `warn` and `allow` findings
/// pass their span through unchanged; `redact` and `block` findings have
/// their span replaced by `formatter`'s output.
///
/// `findings` need not be pre-sorted or non-overlapping; this function
/// establishes both before producing output.
///
/// # Errors
///
/// - [`SecretScanErrorCode::InvalidFindings`] when a finding's range falls
///   outside `input`, is not character-aligned, or overlaps another
///   finding.
/// - [`SecretScanErrorCode::PlaceholderFailure`] when `formatter` fails.
/// - [`SecretScanErrorCode::InvalidPlaceholder`] when `formatter` returns an
///   empty placeholder, a placeholder longer than
///   [`MAX_PLACEHOLDER_LENGTH`], or a placeholder that reproduces any
///   eligible matched value.
///
/// No error carries `input`, a matched value, or a placeholder.
pub fn redact(
    input: &str,
    findings: &[Finding],
    formatter: &dyn PlaceholderFormatter,
) -> Result<String, SecretScanError> {
    let ordered = ordered_and_disjoint(findings, input)?;
    let forbidden = ForbiddenMatchedText::build(input, &ordered);

    let mut output = String::with_capacity(input.len());
    let mut cursor = 0;
    let mut placeholder_index = 0;
    for finding in ordered {
        if !finding.action().replaces_text() {
            continue;
        }

        let range = finding.range();
        output.push_str(&input[cursor..range.start()]);

        placeholder_index += 1;
        let context = PlaceholderContext::new(placeholder_index);
        let placeholder = formatter
            .format(finding, &context)
            .map_err(|_| SecretScanError::new(SecretScanErrorCode::PlaceholderFailure))?;
        if placeholder.is_empty()
            || placeholder.len() > MAX_PLACEHOLDER_LENGTH
            || forbidden.contains(&placeholder)
        {
            return Err(SecretScanErrorCode::InvalidPlaceholder.into());
        }
        output.push_str(&placeholder);

        cursor = range.end();
    }
    output.push_str(&input[cursor..]);
    Ok(output)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::{Action, Confidence};

    fn finding(id: &str, start: usize, end: usize, action: Action) -> Finding {
        Finding::new(
            id,
            "synthetic-credential",
            "synthetic-detector",
            Confidence::High,
            action,
            ByteRange::new(start, end).unwrap(),
        )
        .unwrap()
    }

    #[test]
    fn reconstructs_repeated_and_adjacent_values_in_one_deterministic_pass() {
        let input = "SYNTHETIC_ONE|SYNTHETIC_ONE|SYNTHETIC_TWO";
        let findings = [
            finding("finding-3", 28, input.len(), Action::Block),
            finding("finding-1", 0, 13, Action::Redact),
            finding("finding-2", 14, 27, Action::Redact),
        ];
        let output = redact(input, &findings, &default_placeholder_formatter).unwrap();
        assert_eq!(output, "<SECRET_1>|<SECRET_2>|<SECRET_3>");
        assert_eq!(
            output,
            redact(input, &findings, &default_placeholder_formatter).unwrap()
        );
    }

    #[test]
    fn leaves_warn_and_allow_findings_unchanged_without_consuming_numbers() {
        let input = "WARN|ALLOW|REDACTED_VALUE";
        let findings = [
            finding("finding-1", 0, 4, Action::Warn),
            finding("finding-2", 5, 10, Action::Allow),
            finding("finding-3", 11, input.len(), Action::Redact),
        ];
        assert_eq!(
            redact(input, &findings, &default_placeholder_formatter).unwrap(),
            "WARN|ALLOW|<SECRET_1>"
        );
    }

    #[test]
    fn redacts_adjacent_findings_without_dropping_or_duplicating_text() {
        let input = "SYNTHETIC_ONESYNTHETIC_TWO";
        let findings = [
            finding("finding-1", 0, 13, Action::Redact),
            finding("finding-2", 13, 26, Action::Redact),
        ];
        assert_eq!(
            redact(input, &findings, &default_placeholder_formatter).unwrap(),
            "<SECRET_1><SECRET_2>"
        );
    }

    #[test]
    fn supports_typed_placeholders() {
        let input = "SYNTHETIC_REVOKED_VALUE";
        let findings = [finding("finding-1", 0, input.len(), Action::Redact)];
        assert_eq!(
            redact(input, &findings, &typed_placeholder_formatter).unwrap(),
            "<SYNTHETIC_CREDENTIAL_1>"
        );
    }

    #[test]
    fn sanitizes_formatter_failures() {
        let input = "SYNTHETIC_REVOKED_VALUE";
        let findings = [finding("finding-1", 0, input.len(), Action::Redact)];
        let failing = |_: &Finding, _: &PlaceholderContext| Err(FormatterFailure);
        let error = redact(input, &findings, &failing).unwrap_err();
        assert_eq!(error.code(), SecretScanErrorCode::PlaceholderFailure);
        assert!(!error.to_string().contains(input));
    }

    #[test]
    fn rejects_empty_and_oversized_placeholders() {
        let input = "SYNTHETIC_REVOKED_VALUE";
        let findings = [finding("finding-1", 0, input.len(), Action::Redact)];
        let empty = |_: &Finding, _: &PlaceholderContext| Ok(String::new());
        assert_eq!(
            redact(input, &findings, &empty).unwrap_err().code(),
            SecretScanErrorCode::InvalidPlaceholder
        );
        let oversized =
            |_: &Finding, _: &PlaceholderContext| Ok("x".repeat(MAX_PLACEHOLDER_LENGTH + 1));
        assert_eq!(
            redact(input, &findings, &oversized).unwrap_err().code(),
            SecretScanErrorCode::InvalidPlaceholder
        );
    }

    #[test]
    fn rejects_a_placeholder_containing_any_eligible_matched_value() {
        let input = "SYNTHETIC_ONE|SYNTHETIC_TWO|SYNTHETIC_ONE";
        let findings = [
            finding("finding-1", 0, 13, Action::Redact),
            finding("finding-2", 14, 27, Action::Redact),
            finding("finding-3", 28, input.len(), Action::Redact),
        ];
        let formatter = |_: &Finding, context: &PlaceholderContext| {
            Ok(if context.placeholder_index() == 1 {
                "<SYNTHETIC_TWO>".to_string()
            } else {
                format!("<REMOVED_{}>", context.placeholder_index())
            })
        };
        assert_eq!(
            redact(input, &findings, &formatter).unwrap_err().code(),
            SecretScanErrorCode::InvalidPlaceholder
        );
    }

    #[test]
    fn rejects_formatter_reproduction_of_a_short_caller_supplied_finding() {
        for input in ["x", "xy", "xyz"] {
            let findings = [finding("finding-1", 0, input.len(), Action::Redact)];
            let value = input.to_string();
            let formatter = move |_: &Finding, _: &PlaceholderContext| Ok(format!("<{value}>"));
            assert_eq!(
                redact(input, &findings, &formatter).unwrap_err().code(),
                SecretScanErrorCode::InvalidPlaceholder,
                "{input}"
            );
        }
    }

    #[test]
    fn rejects_overlapping_findings() {
        let input = "SYNTHETIC_REVOKED_VALUE";
        let findings = [
            finding("finding-1", 0, 10, Action::Redact),
            finding("finding-2", 5, 15, Action::Redact),
        ];
        let error = redact(input, &findings, &default_placeholder_formatter).unwrap_err();
        assert_eq!(error.code(), SecretScanErrorCode::InvalidFindings);
        assert!(!error.to_string().contains(input));
    }

    #[test]
    fn rejects_out_of_range_findings() {
        let input = "SYNTHETIC_REVOKED_VALUE";
        let findings = [finding("finding-1", 0, input.len() + 1, Action::Redact)];
        assert_eq!(
            redact(input, &findings, &default_placeholder_formatter)
                .unwrap_err()
                .code(),
            SecretScanErrorCode::InvalidFindings
        );
    }

    #[test]
    fn returns_input_unchanged_when_no_findings_replace_text() {
        let input = "ordinary text";
        assert_eq!(
            redact(input, &[], &default_placeholder_formatter).unwrap(),
            input
        );
    }
}
